### PR TITLE
[STOS-88] Scheduler

### PR DESCRIPTION
**Pull request by AdamNiesiobedzki**

### CODE REVIEW

**Comment by lukrzak** [comment]
[JIRA Ticket](https://tpalinski.atlassian.net/browse/STOS-88)

File: scheduler/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        pass
+
+    @abstractmethod
+    def change_workers_count(self) -> None:
+        pass
+ 
+
+class Scheduler(IScheduler):
+    def __init__(self, initial_workers: int):
+        self.task_queue: queue.Queue[TaskData] = queue.Queue()
```
Comment (tpalinski) [inline scheduler/scheduler.py]: Extract properties for readability?

File: scheduler/scheduler.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+from abc import ABC, abstractmethod
+from typing import override
+import threading
+import queue
+import uuid
+from manager.types import TaskData, Worker
+
+class IScheduler(ABC):
+
```
Comment (lukrzak) [inline scheduler/scheduler.py:12]: Add documentation for interface methods

File: scheduler/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    def __init__(self, initial_workers: int):
+        self.task_queue: queue.Queue[TaskData] = queue.Queue()
+        self.workers_queue: queue.Queue[Worker] = queue.Queue()
+        self.workers_delta: int = 0
+        self.lock: threading.Lock = threading.Lock()
+
+    @override
+    def init_workers(self, number_of_workers: int) -> None:
+        for _ in range(number_of_workers):
+            self.workers_queue.put( Worker(status="awaiting", worker_id=uuid.uuid4()))
```
Comment (lukrzak) [inline scheduler/scheduler.py]: Remove space after put method

File: scheduler/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    @override
+    def register_new_task(self, taskData: TaskData) -> None:
+        self.task_queue.put(taskData)
+        self.manage_workers()
+
+    @override
+    def manage_workers(self) -> None:
+        with self.lock:
```
Comment (tpalinski) [inline scheduler/scheduler.py]: Move locks to helper function

File: worker/worker.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import os
+import shutil
+import time
+
+input_path = os.getenv('INPUT_PATH')
+output_path = os.getenv('OUTPUT_PATH')
+
+if not input_path or not output_path:
+    raise ValueError("Both INPUT_PATH and OUTPUT_PATH environment variables must be set.")
```
Comment (lukrzak) [inline worker/worker.py]: Use `os.environ['ENV_VAR']`. These will raise KeyException if not set

File: worker/dockerfile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.9-slim
```
Comment (lukrzak) [inline worker/dockerfile:1]: I would change it to 3.11 for consistency

**Review by lukrzak — COMMENTED** [review]
In general:
- Write code defensively (type & null checking)
- Write docstrings' to interfaces
- Implement tests

File: scheduler/test/test.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+abcdsadfsafsa
```
Comment (tpalinski) [inline scheduler/test/test.txt]: 👍🏻

**Review by tpalinski — APPROVED** [review]
Great changes, very nice 
download

**Pull request merged**